text/0000-javascript-module-api.md

```diff
@@ -0,0 +1,1058 @@
+[migrator](#migration) on them, then replace the extracted code with the updated
+version. For the API docs, this tool would need to be able to handle Markdown embedded
+in JSDoc-style documentation.
+
+The benefit of this approach is that, once we have verified the script works
+reliably, we can wait until the last possible moment to make the switch. If we
+attempt to update everything by hand, the duration and tediousness of that
+process will likely take out an effective "lock" on the documentation code base,
+where people will put off making big changes because of the potential for merge
+conflicts.
+
+## API Documentation
+
+Our API documentation has long been a source of frustration, because the laundry
+list of (often rarely used or internal) classes makes Ember feel far more
+overwhelming than it really is.
+
+The shift to modules gives us a good opportunity to rethink the presentation of
+our API documentation. Instead of the imposing mono-list, we should group the
+API documentation by package–which, conveniently in this proposal, means they
+will also be grouped by area of functionality.
+
+We should investigate the broader ecosystem to see if there is a good tool that
+generates package-oriented documentation for JavaScript projects. If not, we may
+wish to adapt an existing tool to do so.
+
+## Explaining the Migration
+
+Once the guides and API documentation are updated, modules should be
+straightforward for new learners—indeed, more and more new learners are starting
+with JavaScript modules as the baseline.
+
+The most challenging aspect of teaching the new modules API, counterintuitively,
+will likely be _existing_ users. In particular, for changes that touch nearly
+every file, most teams working on large apps cannot pause work for a week to
+implement the change.
+
+Our focus needs to be:
+
+* Communicating clearly that the existing global build will work for the
+  foreseeable future.
+* Making clear the file size benefits of moving to modules.
+* Building robust tooling that allows even large apps to migrate in a day or
+  two, not a week.
+
+It is important to frame the module transition as a carrot, not a stick. We
+should avoid dire warnings or deprecation notices. Instead, we should provide
+good reporting when doing Ember CLI builds. If an app is compiled in globals
+mode, we can offer suggestions for how to reduce the file size, providing a
+helpful pointer to the modules migration guide. This will make the transition
+feel less like churn and more like an optimization opportunity that developers
+can take advantage of when they have the time or resources.
+
+### Addons
+
+One pitfall is that a _single_ use of the `Ember` global means we have to
+include the entire framework. That means that a developer could migrate their
+entire app to modules, but a single old addon that uses the Ember globals will
+negate the benefits.
+
+This requires a two-pronged strategy:
+
+* Tight integration into Ember CLI
```

Comment:
I feel like we should move this into the RFC as the plan of record for compatibility. This set of permutations demonstrates that the primary audience we need to make sure we address are in the addon community. This is also beneficial as these people are likely more-experienced Ember developers and will be comfortable making those changes.

### Scenario Summary

#### App
1. App written with global Ember? Seems fine, we ship all the things. No clever for you.
2. App written with modules? We can start to do clever things for you at build time to reduce app size or bundle differently.

#### Addons
1. App written with global Ember and includes addon written with global Ember? No change.
2. App written with modules and includes addon written with global Ember? Best attempt at build time codemod the addon to get to keep doing clever things. If it fails the user sets a flag and we stop attempting clever things. This is the least ideal situation.
3. App written with global Ember and includes addon written with modules? Guaranteed safe build time codemod back to global Ember.
4. App written with modules and includes addon written with modules? No work to be done.